### DISCUSSION TITLE
UNION[ ] in response_model( ) chooses the wrong Pydantic model and returns a relationship column data when it shouldn't

### DISCUSSION THREAD

**Discussion by sixth-instinct** [fastapi/fastapi#6826] [Questions] [answered]
Hello everyone,

I'm new to FastAPI but I'm very excited to learn to build an app with this awesome framework. 

I'm trying to build a directory website where users can post several ads (one-to-many relationship), but I'm having an issue with the response being sent out: 

It chooses the wrong Pydantic schema out of the two that I specified in response_model=UNION[Schema_1, Schema_2]. It uses Schema_1 when it should be using Schema_2. I've spent days trying to figure it out to no avail. I don't know which Python extension causes this issue.

Let me show you what I have:

First, I have these 2 database models (I removed most columns to simplify my questions):

```python
class User(Base):
    __tablename__       = 't_users'
    
    id            = Column(Integer, Sequence('seq__t_users__id'), primary_key=True,   index=True) 
    username      = Column(Unicode(20), nullable=False, index=True, unique=True)
    ads           = relationship('Ad', back_populates='author', cascade="all, delete", passive_deletes=True)

class Ad(Base):
    __tablename__      = 't_ads'
    
    id             = Column(Integer, Sequence('seq__t_ads__id'), primary_key=True, index=True)
    author_id      = Column(   Integer, ForeignKey('t_users.id', ondelete='CASCADE'), nullable=False)
    
    title          = Column(Unicode(70), nullable=False)
    description    = Column(UnicodeText(),  nullable=False)
    
    author         = relationship('User', back_populates='ads')
```
I want to be able to retrieve a user from the User table with or without the ads column. So I created this function:

```python
def db_get_one_user_by_id(db_session:Session, user_id:int, should_load_ads_column:bool=False):
    ads_column      = models.User.ads
    ads_load_option = joinedload(ads_column) if should_load_ads_column else lazyload(ads_column)
  
    return db_session\
                    .query(models.User)\
                    .options(ads_load_option)\
                    .filter(models.User.id == user_id)\
                    .one()
```
The function above basically says: If I want to retrieve the user along with his ads, I will use **joinedload()**. If I don't want the ads, only the user, I'll use **lazyload()**. This produces the expected responses:

With ads:
```python
{
  "username": "user_0001",
  "id": 2,
  "ads": [
    {
      "id": 1,
      "title": "my title 2",
      "description": "my desc 2",
      "author_id": 2
    }
  ]
}
```
Without ads:
```python
{
  "username": "user_0001",
  "id": 2,
}
```
**Problem**

However, when I started applying Pydantic models and UNION[ ] to the response_model, an unexpected behavior occurred: The response would still contain the ads like below even though I did not request the data (by using lazyload()) Strange!

With ads even though not requested:
```python
{
  "username": "user_0001",
  "id": 2,
  "ads": [
    {
      "id": 1,
      "title": "my title 2",
      "description": "my desc 2",
      "author_id": 2
    }
  ]
}
```

Here's my router:
```python
@router.get('/id/{user_id}', response_model=Union[schemas.User_LoadAds_Out, schemas.User_Out])
def get_one_user_by_id(user_id:int, should_load_ads_column:bool=False, db_session:Session=Depends(d.get_db_session)):
    return crud.db_get_one_user_by_id(db_session=db_session, user_id=user_id, should_load_ads_column=should_load_ads_column)
```
My Pydantic schemas:
```python

class User_Out(BaseModel):
    class Config:
        orm_mode        = True
        
    id                  : int
    username            : str
    
class Ad_Out(BaseModel):
    class Config:
        orm_mode        = True

    id                  : int
    author_id           : int
    title               : str
    description         : str

class User_LoadAds_Out(BaseModel):
    class Config:
        orm_mode        = True
        
    id                  : int
    username            : str
    ads                 : Optional[List[Ad_Out]] = None
    
```
No matter which query I choose (get user along with his ads OR without them), it will always choose a schema for user with ads (User_LoadAds_Out instead of the correct User_Out schema). Why is that? Why did it choose the wrong schema? Did I do something wrong with my code?

Since I put echo=True in my create_engine( ), I could see the SQL query in my terminal, and it showed that it did not run a query with a JOIN to the Ads table. So where did the Ads information come from?

FastAPI 0.63.0
Pydantic 1.7.3
SQLAlchemy 1.3.23
Uvicorn 0.13.3
psycopg2 2.8.6
Python 3.8.7
PostgreSQL 11.10

Thank you!!

**Comment by ycd** [comment]
See the documentation: https://fastapi.tiangolo.com/tutorial/extra-models/#union-or-anyof

> When defining a Union, include the most specific type first, followed by the less specific type.

**Comment by sixth-instinct** [comment]
Thanks for the reply, ycd.

I did move my schema with the ads column to be before the one without the ads column in my UNION[ ], but it did not help.
```python
response_model=Union[schemas.User_LoadAds_Out, schemas.User_Out]
```
I wish there were a way to assign your response model dynamically through a function.

**Comment by insomnes** [comment, marked as answer]
@sixth-instinct hello!
> I wish there were a way to assign your response model dynamically through a function.

Technically you can create modeled result by yourself:
```python3
from fastapi import FastAPI
from pydantic import BaseModel
from typing import Union


class Simple(BaseModel):
    id: int


class Specific(Simple):
    description: str


RESULT = {
    "id": 1,
    "description": "test"
}

app = FastAPI()


@app.get("/", response_model=Union[Specific, Simple])
def get_main(specific: bool = False):
    if specific:
        result = Specific(**RESULT)
    else:
        result = Simple(**RESULT)

    return result

```

**Comment by sixth-instinct** [comment]
Hello Insomnes,

Yes, your solution works! Thanks!!

Because I'm getting my result from SQLAlchemy ORM operation, I get an object rather than a dictionary that you have in your example, so I would have to covert it into dictionary first as shown below. Is this correct?

```python
return schemas.Ad_Author_Out(**result.__dict__) if should_load_author_column else schemas.Ad_Out(**result.__dict__)
```
What if I want to return a list of ads? The schema class seems to take only a dictionary.

--

I also found a workaround by using noload() instead of lazyload():
```python
 ads_load_option = joinedload(ads_column) if should_load_ads_column else noload(ads_column)
```
If I query for one user and don't want his ads included in the result, noload() would still return the ads column in the result but with a null value (or an empty list [ ] if I query for all( )), while lazyload() would not include the column at all, which I think is cleaner, like your solution. Because the results for a query for user including or excluding their ads have the same data schema (both now have the ads column), I could simply use the same Pydantic model for both.

But I wonder if your solution is better because it produces a cleaner result (no ads column = null or [ ] if the column is not requested). But it seems I would need to convert the database result into a dictionary first and then unpack it before the data can be used for the Pydantic class/model. If I'm retrieving a list of 1000 ads, would your solution be slower than my workaround? Or my workaround would still do the same conversions in the background?

Thanks.

**Comment by adrianrv** [comment]
Hi guys!

@insomnes thanks for your example it works fine, same the one in the documentation @ycd pointed out.

In the other hand I am experience exactly the same issue @sixth-instinct mentioned. The issue can be reproduced when using default value in the `Specific` model, like:

```python
class Specific(Simple):
    description: str = "default"
```

Then, you will see that no matter the value of the `specific` param,  the response always will contain the `Specific` model with the default value in description.

```json
{
    "id": 1,
    "details": "default"
}
```

@sixth-instinct did you try to remove the default value for your `Ads` model?

```python
class User_LoadAds_Out(BaseModel):
    class Config:
        orm_mode = True
        
    id : int
    username: str
    ads: Optional[List[Ad_Out]]
```

**Comment by insomnes** [comment]
@sixth-instinct 
>Because I'm getting my result from SQLAlchemy ORM operation, I get an object rather than a dictionary that you have in your example, so I would have to covert it into dictionary first as shown below. Is this correct?

You can use `orm_mode` 
https://fastapi.tiangolo.com/tutorial/sql-databases/#use-pydantics-orm_mode

>What if I want to return a list of ads? The schema class seems to take only a dictionary.

You can use: `response_model=List[Union[Specific, Simple]]`
Or at python 3.9 i guess `response_model=list[Union[Specific, Simple]]`
And create lists: `result = [Model(item) for item in sql_answer]`

But as @adrianrv mentioned created model still goes through process of validation (and so default value creation) of `response_model` parameter. So i am not sure now, that it is good approach for your problem.

@adrianrv 
>Then, you will see that no matter the value of the specific param, the response always will contain the Specific model with the default value in description.

You can try to hack this with `response_model_exclude_unset` or `response_model_exclude_defaults` parameters in path decorator:
https://fastapi.tiangolo.com/tutorial/response-model/#use-the-response_model_exclude_unset-parameter

**Comment by sixth-instinct** [comment]
Hello @insomnes,

Thanks again for your response.

Your approach to return a list of ads works! This is my code:

Get all ads (with or without the author column data):
```python
@router.get('/all', response_model=List[Union[schemas.Ad_LoadAuthor_Out, schemas.Ad_Out]])
def get_all_ads(should_load_author_column:Optional[bool]=Query(False,alias='load-author'), db_session:Session=Depends(d.get_db_session)):
    result = crud.db_get_all_ads(db_session=db_session, should_load_author_column=should_load_author_column)
    return [schemas.Ad_LoadAuthor_Out(**item.__dict__) for item in result] if should_load_author_column else [schemas.Ad_Out(**item.__dict__) for item in result]
```
The returned result seems correct. It uses the correct Pydantic model. Perhaps because both models don't have any optional field (because ads are children of user, an ad always has a user (author)).

But if I use it on the query to get all users with or without their ads column, it doesn't seem to work. It always include the ads column whether my query requests it or not. It looks like the optional Ads column in the Pydantic model for a query for users along with their ads (I had to set the ads column optional because even though my query requests the column, the user may still not have any ads yet) throws it off. If I removed the Optional option for the ads column, it works as intended. Maybe this is what you mean that this might not be a good approach to my problem. In that case, I'll stick with my workaround using noload(). Thanks for helping me out on this. I appreciate it.

Thanks.

**Comment by sixth-instinct** [comment]
Hello @adrianrv,

Thanks for your feedback.

I removed all the default values in my Pydantic models, but somehow I still got the same results: the ads column with its content is always included in the response whether my db query asks for it or not. Maybe I'm missing something. I think I'm just going to use my noload() workaround. Thanks for your help. I really appreciate it.

Thanks.

**Comment by holyoaks** [comment]
I don't think this should be closed. IMO, having to do all sorts of workarounds or order your response models in a particular way is too, too fragile to suffice. FastAPI, knows what Type of object it has to serialize, but it throws away that Typing information in `routing.serialize_response`, before validating the `response_content`. The validation then grabs the first pydantic model it finds, which may not be the correct one. Why would we throw away the typing information before validation?

My example, is a (very simplified) endpoint that returns a List of one or more Pydantic models, all of them subclasses of one base Pydantic model.

``` python
class BaseClass(BaseModel):
    id: int
    name: str

class SubclassA(BaseClass):
    level: str
    previous_level: str

class SubclassB(BaseClass):
    email: str
    subject: str
    ...

class SubclassC(BaseClass):
    level: str
    previous_level: Optional[str]
   ...

class SubclassD(BaseClass):
    level: str
    next_level: str
    ...

async def get_classes() -> List[Union[SubclassA, SubclassB, SubclassC, SubclassD]]:
   ...
```

SubclassA and SubclassC both have a `level` key, and what happens is, FastAPI converts an instance of SubclassC into an instance of SubclassA and throws away any additional keys, simply because the SubclassA happens to share a key with SubclassC. Worse, in my example, is by throwing away the additional keys, the response is IMPROPERLY validating. SubclassA requires `previous_level`, but if a SubclassC instance that does not require that key is instead validated against SubclassA, a validation error will occur _even though there is no validation error_. I understand that the current "solution" is to list SubclassC first, but that doesn't solve the issue, as I may have a SubclassD which would fail validation. 

This approach of relying on order, IMO, is too fragile and is a bug, and this issue should not be closed simply because there is a workaround. IMO, FastAPI should not throw away typing information before validation; it should USE the typing information to properly validate the `response_content`. It is not always feasible to define the "most specific type first", as it is not always clear what constitutes the "most specific type".

**Reply by YuriiMotov** [reply]
I see a lot of reactions to this message. Let's try to figure out with it.

The following test (in details) works fine with Pydantic V2.

<details>

```python
from typing import Optional, Union
from fastapi import FastAPI
from pydantic import BaseModel
from fastapi.testclient import TestClient


class BaseClass(BaseModel):
    id: int
    name: str


class SubclassA(BaseClass):
    level: str
    previous_level: str


class SubclassB(BaseClass):
    email: str
    subject: str


class SubclassC(BaseClass):
    level: str
    previous_level: Optional[str]
    additional_field: str


class SubclassD(BaseClass):
    level: str
    next_level: str


app = FastAPI()


RES = [
    SubclassA(id=1, name="A1", level="L1", previous_level="L0"),
    SubclassB(id=2, name="B1", email="example@example.com", subject="Subject 1"),
    SubclassC(
        id=3, name="C1", level="L2", previous_level=None, additional_field="Extra Info"
    ),
    SubclassD(id=4, name="D1", level="L3", next_level="L4"),
    SubclassC(
        id=5,
        name="C1",
        level="L2",
        previous_level="L1",
        additional_field="Extra Info 2",
    ),
]


@app.get("/classes")
async def get_classes() -> list[Union[SubclassA, SubclassB, SubclassC, SubclassD]]:
    return RES


def test():
    client = TestClient(app)

    response = client.get("/classes")
    assert response.status_code == 200
    data = response.json()

    assert len(data) == 5
    assert data == [
        {"id": 1, "name": "A1", "level": "L1", "previous_level": "L0"},
        {"id": 2, "name": "B1", "email": "example@example.com", "subject": "Subject 1"},
        {
            "id": 3,
            "name": "C1",
            "level": "L2",
            "previous_level": None,
            "additional_field": "Extra Info",
        },
        {"id": 4, "name": "D1", "level": "L3", "next_level": "L4"},
        {
            "id": 5,
            "name": "C1",
            "level": "L2",
            "previous_level": "L1",
            "additional_field": "Extra Info 2",
        },
    ]
```

</details>

It fails with Pydantic V1 (`additional_field` of the last item is dropped), but it can easily be fixed by adding `extra = Extra.forbid` to `BaseClass`.

```python
class BaseClass(BaseModel):
    id: int
    name: str

    class Config:
        extra = Extra.forbid
```

Some of the mentioned problems were probably fixed since the time you posted it. Although, FastAPI still doesn't take into account actual types of objects you return from the endpoint (not an issue, this is how it works).

If you still need more control over this, you can [return Response directly](https://fastapi.tiangolo.com/advanced/response-directly/) to bypass internal validation.

**Comment by bhexblah** [comment]
I would be happy to have at least a workaround on using generics.

```
from fastapi import FastAPI
from pydantic import BaseModel
from pydantic.generics import GenericModel
from typing import List, Union, Generic, TypeVar


class MyBase(BaseModel):
    id: int

class MySub1(MyBase):
    x: int = 1
    y: str = "foo"

class MySub2(MyBase):
    x: int = 2
    z: str = "bar"

MyBaseT = TypeVar('MyBaseT', bound=MyBase)

class MyList(GenericModel, Generic[MyBaseT]):
    items: List[MyBaseT]
    description: str

app = FastAPI()

@app.get("/", response_model=Union[MyList[MySub1], MyList[MySub2]])
def get_main(sub: int = 1):
    if sub == 1:
        clz = MySub1
        data = {"items": [{"id": 1, "x": 5, "y": "aaa"}], "description": ""}
    else:
        clz = MySub2
        data = {"items": [{"id": 1, "x": 5, "z": "bbb"}], "description": "does not work"}

    print(MyList[clz](**data))
    return MyList[clz](**data)
```
```
http://<host>:<port>/?sub=1
# json
{"items":[{"id":1,"x":5,"y":"aaa"}],"description":""}
# print
items=[MySub1(id=1, x=5, y='aaa')] description=''
```
```
http://<host>:<port>/?sub=2
# json
{"items":[{"id":1,"x":5,"y":"foo"}],"description":"does not work"}
# print
items=[MySub2(id=1, x=5, z='bbb')] description='does not work'
```

**Comment by Prometheus3375** [comment]
Another example of the issue:

```python
from typing import Literal

from fastapi import FastAPI
from pydantic import BaseModel


class Result1(BaseModel):
    result: str = 'result'
    connected_to_connected: str | None = None


class Result2(BaseModel):
    result: str = 'result'
    connected_matched: str | None = None
    connected_to_connected_matched: str | None = None


class Result3(BaseModel):
    some_cool_result: str = 'some_cool_result'


app = FastAPI()


def foo(value: int, /) -> Result1 | Result2 | Result3:
    match value:
        case 1:
            return Result1(connected_to_connected='connected_to_connected')
        case 2:
            return Result2(
                connected_matched='connected_matched',
                connected_to_connected_matched='connected_to_connected_matched',
                )
        case 3:
            return Result3()


@app.get('/{value}', response_model_exclude_none=True)
def api1(value: Literal['1', '2', '3']) -> Result1 | Result2 | Result3:
    return foo(int(value))


@app.get('/clean/{value}', response_model=None, response_model_exclude_none=True)
def api2(value: Literal['1', '2', '3']) -> Result1 | Result2 | Result3:
    return foo(int(value))
```

Both `/2` and `/3` return `{"result":"result"}`
`/clean/2` returns `{"result":"result","connected_matched":"connected_matched","connected_to_connected_matched":"connected_to_connected_matched"}`
`/clean/3` returns `{"some_cool_result":"some_cool_result"}`
Both endpoints for `1` return expected result.

Essentially, if response model is a union of models, then only the first one is used. I expect FastApi to validate that the result matches one of the models, but instead the result is just cast to `Result1`.

---

In addition, I noticed that the docs about [unions since Python 3.10](https://fastapi.tiangolo.com/tutorial/extra-models/#union-in-python-310) are incorrect. `|` is a valid operation on types. I also passed `Result1 | Result2 | Result3` to `response_model` argument of `api1` and no error was emitted. With such argument passed, I got identical results for the previous test.

That whole paragraph should be removed as is it misleading.

**Reply by YuriiMotov** [reply]
To understand the problem, try converting all your results to dict and then validate this dict using your `Union` type.

With Pydantic V1 FastAPI tries to validate data using all types of dict from left to right and the first that passes validation is used.
Pydantic V2 is smarter and it validates results correctly even without `extra=forbid` (it tries to find the best fit, not just first that passes validation).

Adding `extra = Extra.forbid` to all models solves mentioned problems.
